Handle default values

lars-petter-hauge opened the pull request:
Starting out with some tests.

Some of this is even pseudocode, so it's more to get a foundation for discussions how one would expect defaults to behave

lars-petter-hauge commented:
Some points of discussion:
 - Should it be top-down, or bottom-up? The `test_default_presedence` is currently set top-down
 - If it is top-down, how do you handle partial defaults? `test_partial_default_values` removes default for one of the keys in the top level. The behaviour asserted in the test is that it still uses whatever is defined in the top-default.
 - Should there be issued warning for redundant defaults in a schema?

lars-petter-hauge commented on configsuite/schema.py:
```diff
@@ -72,11 +78,19 @@ def assert_valid_schema(schema):
+    if MK.Default in schema:
+        if not allow_default:
+            fmt = "Default value is only allowed for contents in NamedDict"
+            raise ValueError(fmt)
```
I would really like to add the name of the schema that led to this error, unsure how to do this without passing another argument to the function.. quickly becomes lots of minor arguments here

markusdregi replied:
Let us leave it for now. The ease of this validation is not the most importance as it should never hit anyone but the developer. But, when that is said, passing down the key path similarly to in the data validation is definitively possible and useful. Suggest we create an issue for this and implement it broadly in the schema validation if it is to carried out ;)

lars-petter-hauge commented on tests/test_default_values.py:
```diff
@@ -35,3 +37,72 @@ def test_default_values(self):
+        ] = "incorrect default type"
+        with self.assertRaises(ValueError):
+            configsuite.ConfigSuite(raw_config, car_schema)
+            # self.assertTrue(msg.exception.startswidth("default problems!"))
```
Would like to assert the message to be sure it's the correct exception that is caught and not another `ValueError` somewhere else. `unittest2` has this functionality, but unfortunately not `unittest`...

markusdregi replied:
Perhaps not as nice, but:

```python
with self.assertRaises(ValueError) as error_context:
    configsuite.ConfigSuite(raw_config, car_schema)

self.assertTrue(error_context.exception.startswidth("default problems!"))
```
should work... We can of course also consider dragging in `unittest2`, but I would assume that this will do for now?

lars-petter-hauge replied:
This is how it's done in `unittest2` as well. I tried that, somehow I managed to mess it up.. Apparently it's backported to unittest, so I'll remove the comments and have them execute!

lars-petter-hauge replied:
```
        with self.assertRaises(ValueError) as msg:
            configsuite.ConfigSuite(raw_config, car_schema)
>       self.assertTrue(msg.exception.startswidth("default problems!"))

E       AttributeError: 'exceptions.ValueError' object has no attribute 'startswidth'
```
Yeh.. this is what I tested the first time.. am I missing something?

lars-petter-hauge replied:
Right.. Damn it! alright... I'll fix `str` it!

lars-petter-hauge commented on configsuite/meta_keys.py:
```diff
@@ -17,10 +17,7 @@
 """
 
 
-import enum
```
I'll add a better commit message on this one

lars-petter-hauge commented:
Alright, so I've got a few more things to add on this one - but a first implementation is close to ready at least!

Need to add a test where the schema only contains a single entry with a basictype

lars-petter-hauge commented on configsuite/config.py:
```diff
@@ -26,6 +26,13 @@
 from .meta_keys import MetaKeys as MK
 
 
+_RETURN_TYPE = {
```
Not the best solution this one, open for other suggestions..

Considered including the empty container as part of the `Type` - `collections.NamedTuple`, but didn't really fan out

lars-petter-hauge commented on configsuite/config.py:
```diff
@@ -300,9 +318,9 @@ def _build_dict_snapshot(self, config, schema):
+    def _build_snapshot(self, config, schema, key_type=None):
         if config is None:
-            return None
+            return _RETURN_TYPE.get(key_type.name)
```
This is an obvious miss if the `config is `None` and the `key_type` has default value

lars-petter-hauge commented on setup.py:
```diff
@@ -63,10 +63,7 @@ def _load_readme():
-        "setuptools_scm",
-        "setuptools_scm_about",
-    ],
+    setup_requires=["setuptools_scm", "setuptools_scm_about"],
```
`black` insisted on this one, although I believe it was `black` who formatted the previous setup.py as well.. Could there be version differences?

lars-petter-hauge commented:
>Need to add a test where the schema only contains a single entry with a basictype

Considering that default values should only be applied for `NamedDicts` then we don't need this

lars-petter-hauge commented:
The commit `Validate Default type` is unfinished. Either it will be completed in this PR, or we'll state that we do not validate defaults and then fix this later.

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,108 @@
+        self.assertTrue(config_suite.valid)
+
+        # Unspecified containers are defaulted to empty container of their type
+        self.assertEqual(config_suite.snapshot.owner, {})
```
I would have expected this to be an empty `tuple`...

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,108 @@
+        # Unspecified containers are defaulted to empty container of their type
+        self.assertEqual(config_suite.snapshot.owner, {})
+        self.assertEqual(config_suite.snapshot.color, {})
+        self.assertEqual(config_suite.snapshot.minor_incidents, [])
```
Same with this one...

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,108 @@
+
+        self.assertTrue(config_suite.valid)
+
+        # Unspecified containers are defaulted to empty container of their type
```
Please verify that `config_suite.tire.rim` and `config_suite.tire.color` is defaulted as expected.

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,108 @@
+
+        car_schema = car.build_schema()
+        # Dict does not allow default
+        car_schema[MK.Content]["owner"][MK.Content][MK.Key][
```
Are not this and the line below doing the exact same thing? Also, I suggest we use a somewhat reasonable default value instead, like `{"default_key": "default_value"}`.

markusdregi replied:
When I look at this, does not the section below test exactly this and hence this block being a bit mute?

markusdregi replied:
Also, please make a similar test for `list`.

markusdregi replied:
And a test that tries to specify a default value for a `BasicType` inside a list or dict, which should not be allowed.

lars-petter-hauge replied:
This is what I want to test:
Given the setup:
```
"owner": {
                MK.Type: types.Dict,
                MK.Required: False,
                MK.Content: {
                    MK.Key: {MK.Type: types.String},
                    MK.Value: {MK.Type: types.Number},
                },
```
I want to update the key and value with default values (which is not allowed: for dicts)

```
"owner": {
                MK.Type: types.Dict,
                MK.Required: False,
                MK.Content: {
                    MK.Key: {MK.Type: types.String, MK.Default="correct default type"},
                    MK.Value: {MK.Type: types.Number, MK.Default=10},
                },
```
In that case the lines are updating key and value separately.

The list test is tested before the dict.

These tests does indeed set the values to `BasicType` for each of list and dict.

Afterwards I'm setting up a test where the `MK.Default` is set on the `NamedDict` itself
```
# Default only valid at basic types
        schema_value_content = {
            MK.Type: types.NamedDict,
            MK.Default: {"profile": 200, "width": 20},
            MK.Content: {
                "profile": {MK.Type: types.Number},
                "width": {MK.Type: types.Number},
            },
        }
```

markusdregi commented on tests/data/car.py:
```diff
@@ -0,0 +1,71 @@
+from configsuite import types
+
+
+def build_schema():
```
I would really like to test the `Default`-values also for `NamedDicts` inside lists and dicts. As this is the real purpose of them in the schema anyway. You can either extent this schema, or try to find a another case in the test data that already features this...

lars-petter-hauge replied:
Extended the schema, but I really don't have any strong preferences.. I didn't want to change existing schemas when I made the PR, but it's possible to make use of some of the other ones.. If we find that there's too many data sets then I'll happily put the same setup in some existing schema

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,164 @@
+        }
+        car_schema[MK.Content]["tire"][MK.Content]["dimension"] = schema_value_content
+
+        with self.assertRaises(ValueError) as msg:
```
Please call the exception wrapper something else then `msg`.

lars-petter-hauge replied:
will fix

markusdregi commented on configsuite/schema.py:
```diff
@@ -32,13 +32,28 @@ def _context_basic_only(elem):
 _META_SCHEMA = {
     MK.Type: types.NamedDict,
-    MK.ElementValidators: (_context_basic_only,),
+    MK.ElementValidators: (_context_basic_only, _correct_default_type),
```
This is more tricky then I first expected. I suggest that we leave this for now (removing the entire `_correct_default_type`) and then, if we are to have a look at this we should do it for all layers, not only the default values in the schema.

lars-petter-hauge replied:
Sure, I think this PR will have plenty enough to handle on its own

lars-petter-hauge commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,164 @@
+
+        # Unspecified containers are defaulted to empty container of their type
+        self.assertEqual(config_suite.snapshot.owner, {})
+        self.assertEqual(config_suite.snapshot.color, {})
```
>  I would expect tire at this point to contain two entries, dimension defaulted to 17 and rim defaulted to green as a namedtuple.

Not sure If I agree.. How should the behaviour be if there's another entry for `tire` that is in fact required?

```
            "tire": {
                MK.Type: types.NamedDict,
                MK.Required: False,
                MK.Content: {
                    "dimension": {
                        MK.Type: types.Number,
                        MK.Default: 17,
                        MK.Required: False,
                    },
                    "rim": {
                        MK.Type: types.String,
                        MK.Default: "green",
                        MK.Required: False,
                    },
                    "required_entry: {
                        MK.Type: types.String,
                    },
                },
            },
```
Whenever we allow for defaulting the upper levels (which the `test_invalid_defaults` checks is _not_ allowed), then we can return something else.

lars-petter-hauge replied:
Based on your response above I suppose it would be `None`

lars-petter-hauge replied:
>I would expect color to contain None both for interior and chassis, not to be a dictionary.

So in the case of a vast schema containing only `NamedDict`, nothing is required and there are no defaults, you would end up with every entry given, and all containing a `None` value?

markusdregi replied:
That is exactly what I mean, yes. No collection should be defaulted to None, but empty versions of itself instead. For a name dict without a default, the value should be defaulted to None, as today. But, if a required value is not specified, I would expect a MissingKeyError in the same way as one would currently get...

lars-petter-hauge replied:
>But, if a required value is not specified, I would expect a MissingKeyError in the same way as one would currently get...

of course..

lars-petter-hauge commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,164 @@
+        self.assertEqual(config_suite.snapshot.tire.rim, "green")
+
+        # Named dict within dict
+        extracted_owner = {key: value for key, value in config_suite.snapshot.owner}
```
Yepp, that looks better (and scales better)

lars-petter-hauge commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,164 @@
+        self.assertEqual(config_suite.snapshot.incidents[1].location, "somewhere else")
+        self.assertEqual(config_suite.snapshot.incidents[1].casualties, 1)
+
+    def test_partial_default_values(self):
```
Yes, this isn't needed anymore

lars-petter-hauge commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,164 @@
+        self.assertEqual(config_suite.snapshot.tire.rim, "green")
+        self.assertEqual(config_suite.snapshot.tire.dimension, 15)
+
+    def test_invalid_defaults(self):
```
will fix

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,187 @@
+        for (expected_name, expected_location), config_owner in zip(
+            owner_data, sorted(config_suite.snapshot.owner)
+        ):
+            self.assertEqual(expected_name, config_owner.value.name)
```
All asserts in config suite are written with expected value first, like your test here. Please make that consistent in this test also...

lars-petter-hauge replied:
consistensifying it!

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,187 @@
+        ] = car._inch_to_cm
+        config_suite = configsuite.ConfigSuite(raw_config, car_schema)
+        self.assertTrue(config_suite.valid)
+        self.assertEqual(config_suite.snapshot.tire.dimension, 17 * 2.54)
```
`assertAlmostEqual`?

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,187 @@
+    def test_transformations_applied(self):
+        raw_config = car.build_all_default_config()
+        car_schema = car.build_schema()
+        car_schema[MK.Content]["tire"][MK.Content]["dimension"][
```
Would this be easier to read?

```python
tire_schema = car_schema[MK.Content]["tire"]
tire_schema[MK.Content]["dimension"][MK.Transformation] = car._inch_to_cm
```

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,187 @@
+            configsuite.ConfigSuite(raw_config, car_schema)
+        self.assertTrue(str(error_context.exception).startswith("Is default value"))
+
+    def test_invalid_default_type(self):
```
Do we still expect this failure though? Are we doing any default value validation?

lars-petter-hauge replied:
I'll remove the test and whatever remains of the type validation

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,187 @@
+            configsuite.ConfigSuite(raw_config, car_schema)
+        self.assertTrue(str(error_context.exception).startswith("Default value is"))
+
+        # Can't default basic types in lists
```
Please split this in a separate test

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,187 @@
+            configsuite.ConfigSuite(raw_config, car_schema)
+        self.assertTrue(str(error_context.exception).startswith("Default value is"))
+
+        # Can't default basic types in dicts
```
Please split this in a separate test

markusdregi commented on tests/data/car.py:
```diff
@@ -0,0 +1,113 @@
+
+
+@configsuite.transformation_msg("Convert to cm")
+def _inch_to_cm(l):
```
Please remove the prefix `_`, as you are planning to access it outside of this submodule.

markusdregi commented:
Had some formatting comments on the tests, but I agree with the behavior now and think that we are soon ready to move over to the implementation.

Only tests missing are the ones for elements with validators, ensuring that the validators are indeed applied on the final result. For this you need to create a default that is not valid, add a `layer`, `element` and `context`-validator in three separate tests, not overwrite the default so that it "becomes active" and assert that validation fails. Note that it can be the same validator, with the context one ignoring the context all together.

Then, you should also test that transformators are applied as expected. Again, set a default value that is not valid. For instance the string `"123"` when an integer is expected, then add in two different tests, a `Tranformator` and a `ContextTransformator`, both just trying to cast the value to an integer, and then ensure that it is the default value that ends up in the configuration at last.

lars-petter-hauge commented:
Thanks for the tips on the tests, haven't applied anything else than the context validators before.. Working on those right now

lars-petter-hauge commented:
Most of the tests fail now, and they will, until the behaviour of an empty NamedDict is correctly implemented.

Still got some work to do

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,214 @@
+        raw_config = car.build_all_default_config()
+        car_schema = car.build_schema()
+        tire_schema = car_schema[MK.Content]["tire"]
+        tire_schema[MK.Content]["dimension"][MK.ContextTransformation] = car.inch_to_cm
```
This will fail due to the context being passed as an argument. You need to wrap the `inch_to_cm` in another transformator taking the context, but ignoring it. I think I prefer the explicit second version over making `inch_to_cm` take `*args` to facilitate future refactoring. Something about explicit over implicit and those things.... ;)

lars-petter-hauge replied:
Right, true.. I'll think of something else, I agree with being explicit..

markusdregi replied:
I think something like this should do the trick:

```python
@transformation_msg("Transforms from inches to cm with an ignored context")
def inch_to_cm_context(l, _):
    return inch_to_cm(l)
```

No need to change the rest of the test, just swap `inch_to_cm` with `inch_to_cm_context`...

lars-petter-hauge replied:
Yeh, I know..

But I often look at the tests to see how things are supposed to function, and the more tests that actually reflect that the better.. That was the philosophy behind why I changed it to behave as intended.

There's a balance though of course.. Tests should be made to test only one thing at the time..

lars-petter-hauge replied:
But to be honest, the test became cluttered and wasn't really that helpful :P

markusdregi commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,214 @@
+
+        self.assertFalse(config_suite.valid)
+        for err in config_suite.errors:
+            self.assertTrue(err.msg.startswith("Is x a valid date"))
```
Can you also check the type of the errors whenever there are errors?

lars-petter-hauge replied:
I sure can

markusdregi commented:
:tada: The tests are looking good now :tada:

markusdregi commented on configsuite/schema.py:
```diff
@@ -57,7 +58,12 @@ def _context_basic_only(elem):
 
 def assert_valid_schema(schema):
-    _assert_valid_schema_level(schema)
+    allow_default = MK.Type in schema and isinstance(schema[MK.Type], types.BasicType)
```
I guess this should always be `False` since we only allow for defaults inside a `NamedDict`?

lars-petter-hauge replied:
Yes, it's changed a bit, I'll revert this back slightly

markusdregi commented on configsuite/schema.py:
```diff
@@ -72,11 +78,19 @@ def assert_valid_schema(schema):
 
 
-def _assert_valid_schema_level(schema):
+def _assert_valid_schema_level(schema, allow_default=True):
```
This should not be an optional argument?

lars-petter-hauge replied:
fixed

markusdregi commented on configsuite/types.py:
```diff
@@ -187,6 +192,8 @@ def _is_datetime(x):
 Bool = BasicType("bool", _is_bool)
 Date = BasicType("date", _is_date)
 DateTime = BasicType("datetime", _is_datetime)
+Anytype = BasicType("wildcard", _is_anytype)
```
Please make the name of the basic type and its name correspond (`Anytype` vs `wildcard`). Also, I think we should make these private at some point. Together with `Type` and `Callable`, but that is for another day ;)

lars-petter-hauge replied:
fixed

markusdregi commented on configsuite/config.py:
```diff
@@ -300,9 +318,9 @@ def _build_dict_snapshot(self, config, schema):
         )
 
-    def _build_snapshot(self, config, schema):
+    def _build_snapshot(self, config, schema, parent_type=None):
```
Should not be necessary to touch the `snapshot` code. All the validation happens prior to this. The remaining changes in this PR should hit the `_build_initial_merged_config` as far as I can see.

lars-petter-hauge replied:
With the tests we have now it will be impossible to have the implementation in the `_build_snapshot` ;)

I barely started on trying to set this up in the `_build_initial_merged_config`, but encountered a few issues.. Didn't push the changes though, and I didn't get to the bottom of what was wrong before I had to go.. I'll be back on more on this!

lars-petter-hauge commented:
Given that the tests are fine, I'll do rebase now

lars-petter-hauge commented on configsuite/config.py:
```diff
@@ -202,42 +209,61 @@ def _build_transformed_layers(self):
+        all_keys = (key for layer in layers for key in layer.keys())
+        for key in all_keys:
+            child_layers = tuple([layer[key] for layer in layers if key in layer])
+            if len(child_layers) == 0:
```
Will this ever happen? Given that the `all_keys` are aggregated from each layers `keys`?

lars-petter-hauge commented:
First test implementation in https://github.com/equinor/configsuite/pull/70/commits/2d393a2e47ab933df0cd04decbc705d4bfa6b211

Breaking all the tests now, so something's changing! Validation breaks, not unexpected..

`_build_initial_merged_config` will create a config that behaves according to the tests at least.

lars-petter-hauge commented on configsuite/config.py:
```diff
@@ -202,42 +202,70 @@ def _build_transformed_layers(self):
+                for key, val in content_schema.items()
+                if "default" in val or val[MK.Type] == configsuite.types.NamedDict
+            )
+        )
```
This doesn't look good, needs some refactoring

lars-petter-hauge commented:
I believe we had some duplicate test data structure as well, so I've removed the `color` section, and rather did the very small change of popping the `MK.Default` for one of them in the `test_untouched_basic_types`.

Also added a `BasicType` as a first entry

lars-petter-hauge commented:
I'll remove the `class enum to object` commit by the way. First off it doesn't belong in this PR, but foremost, the only reason for doing the change is to have `pprint` work - and that isn't a good reason. Using `enum` gives some context of what it's supposed to do, while changing it to an arbitrary `object` makes it less obvious of it's use.

lars-petter-hauge commented:
This PR also needs to add documentation about usage of default values

lars-petter-hauge commented:
> Starting to look good rocket
> 
> I would like to see a test that verifies that default values of a named dict inside a named dict works. If you pop the `tire` from the `default_config` and verify that you still get `17` and `green` out again?

This is what the `test_empty_named_dict_default_values` does, it uses the `build_all_default_config`

lars-petter-hauge commented on configsuite/schema.py:
```diff
@@ -66,8 +88,8 @@ def _check_allownone_required(schema_level):
 
-def assert_valid_schema(schema):
-    _assert_valid_schema_level(schema)
+def assert_valid_schema(schema, allow_default=False):
```
I believe we discussed earlier that the following is True:
> So in the case of a vast schema containing only NamedDict, nothing is required and there are no defaults, you would end up with every entry given, and all containing a None value?

In that case, why is it not valid to have default values for an item in a `NamedDict` within a `NamedDict` even though no other items have been declared?

This is pretty much what `test_empty_named_dict_default_values` does, as the top level is a `NamedDict` and the `tire` key is another `NamedDict`, and with nothing but `production_date` specified, we still expect the default values for `tire`.

Or is there something I'm missing?

markusdregi replied:
> In that case, why is it not valid to have default values for an item in a NamedDict within a NamedDict even though no other items have been declared?

That is valid, but not to have a default value for a `NamedDict` inside a `NamedDict`. I.e 

```python
MK.Type: types.NamedDict,
MK.Content: {
    "some_key": {
       MK.Type: types.NamedDict,
       MK.Defaulf: {"another_key": 12},
       MK.Content: {
           "another_key": {MK.Type: types.Integer},
       },
   },
}
```

lars-petter-hauge replied:
We have a test for this specific event `test_not_allow_default_for_named_dict`, it is not allowed.

lars-petter-hauge replied:
It is validated in the 

```
@configsuite.validator_msg("Default can only be used for BasicType")
def _check_default_type(schema_level):
    if MK.Default in schema_level:
        return isinstance(schema_level[MK.Type], types.BasicType)
    return True
```

lars-petter-hauge commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,241 @@
+            configsuite.ConfigSuite(raw_config, car_schema)
+        self.assertTrue(str(error_context.exception).startswith("Default value is"))
+
+    def test_invalid_defaults_in_dict(self):
```
Sure thing, and yes I think it's reasonable to split it!

lars-petter-hauge commented on tests/test_default_values.py:
```diff
@@ -0,0 +1,241 @@
+            configsuite.ConfigSuite(raw_config, car_schema)
+        self.assertTrue(str(error_context.exception).startswith("Default value is"))
+
+    def test_invalid_defaults_in_list(self):
```
Will do!

lars-petter-hauge commented:
(wasn't able to reply under your answer..)

>Why did we do this again?

Funny thing is, I wrote a comment here that stated pretty much:

> I've removed the `enum to object` commit as the reason for doing the change was to have `pprint` work nicely. However, the reason is poor and having `enum` provides some context as to how it's going to be used

Then I deleted the commit, and then the default tests failed :sweat_smile: 

So I think it started out with that simple reason, and somehow the implementation change allowed the rest of the PR to function properly..

lars-petter-hauge commented:
It is apparently becoming too complex!

> /home/travis/build/equinor/configsuite/configsuite/schema.py:107:1: C901 'assert_valid_schema_level' is too complex (12)

lars-petter-hauge commented on configsuite/types.py:
```diff
@@ -186,6 +191,8 @@ def _is_datetime(x):
 Bool = BasicType("bool", _is_bool)
 Date = BasicType("date", _is_date)
 DateTime = BasicType("datetime", _is_datetime)
+Anytype = BasicType("anytype", _is_anytype)
```
Agreed!

markusdregi commented:
> This is what the test_empty_named_dict_default_values does, it uses the build_all_default_config

No, because the default config contains `"tire": {"dimension": 15},`. I want you to pop that key and still get the same behaviour.

lars-petter-hauge commented:
```
def build_all_default_config():
    return {"production_date": datetime.datetime(2000, 1, 1)}
```

There's only the `production_date` that has been specified.

It's done like this instead of popping all of the other ones..

markusdregi commented:
> It is apparently becoming too complex!

The complexity bound is specified it in the `.flake8`-file. Just increase it for now ;)

lars-petter-hauge commented:
Increasing complexity! :laughing:

markusdregi commented:
> There's only the production_date that has been specified.

I yield! Was looking at the other config...

markusdregi commented on configsuite/schema.py:
```diff
@@ -25,6 +25,14 @@
+    return True
+
+
+Anytype = types.BasicType("anytype", _is_anytype)
```
`_Anytype`?

lars-petter-hauge replied:
Sure

markusdregi commented on configsuite/schema.py:
```diff
@@ -82,11 +112,19 @@ def assert_valid_schema(schema):
+        if not allow_default:
+            fmt = "Default value is only allowed for contents in NamedDict"
+            raise ValueError(fmt)
+        if not isinstance(level_schema[MK.Type], types.BasicType):
```
This check is also checked by the level schema. I suggest you remove this check ;)

lars-petter-hauge replied:
Yeppsi! And that made the error message better as well

markusdregi commented on docs/source/user_guide/getting_started.rst:
```diff
@@ -569,24 +569,80 @@ value to something else then ``None``, we refer the reader to the next section.
+has the advantage of being able to provide default values for ``BasicTypes``
+within containers. The disadvantage is that you would need to edit the code to
+change the default values and hence site or project specific defaults are not
+suited for this purpose. The second, and currently only implemented, way of
```
>The second, ~~and currently only implemented,~~ way of

lars-petter-hauge replied:
Thanks!

markusdregi commented:
I think the codacy complaints will go away if you rebase on master...

lars-petter-hauge commented:
Did that yesterday ;)

I can't see how the latest commits on master would fix codacy? I see the flake8 config, but nothing about codacy?

lars-petter-hauge commented:
Alright, so I've removed the `enum to object` commit, reasoning stands:

> reason for doing the change was to have pprint work nicely. However, the reason is poor and having enum provides some context as to how it's going to be used

Fixed the code to work with the prior solution, small change in  `if "default" in val`  to `if MK.Default in val` (which also looks better, although it could probably have been this way the first time around).

I've rebased and squashed commits, I believe I'm happy with the PR now

markusdregi commented:
> I can't see how the latest commits on master would fix codacy? I see the flake8 config, but nothing about codacy?

Codacy is running `flake8`, so then the `.flake8` matters ;)
